File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

'''延續上週'''
# 使用isSameTree（recursion）
class Solution:
    def isSubtree(self, root:Optional[TreeNode], subRoot:Optional[TreeNode]) -> bool:
        if not subRoot: # None算任何tree的subtree
            return True
        if not root and subRoot: # and subRoot也寫出來更清楚
            return False
        if self.isSameTree(root, subRoot):
            return True
        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)
    # 也可以寫進inner function（封裝性）
    def isSameTree(self, tree1:Optional[TreeNode], tree2:Optional[TreeNode]) -> bool:
        if not tree1 and not tree2:
            return True
        if tree1 and tree2 and tree1.val == tree2.val:
            return self.isSameTree(tree1.left, tree2.left) and self.isSameTree(tree1.right, tree2.right)
        else:
            return False


        

File: Data Structures & Algorithms/invert-a-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 
class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        if (root == nullptr) return nullptr;

        // 遞迴反轉左右子樹，並將結果保存下來
        TreeNode* left = invertTree(root->left);
        TreeNode* right = invertTree(root->right);
        
        // 交換左右子節點（原地修改，不創建新節點省空間）
        root->left = right;
        root->right = left;
        
        return root;
    }
};
